### PR TITLE
Add Unit test to cover New storage Context in different RMprofile and SMProfile combination

### PR DESCRIPTION
**Pull request by blueww**
## Description
Add  Unit test to cover New storage Context Test in different RMprofile and SMProfile combination.

We have a severe regression issue in the latest stage of release 4.0.0, which is not caught by Unit test.
https://github.com/Azure/azure-powershell/issues/3939

Mark has add test coverage for this issue in https://github.com/Azure/azure-powershell/pull/3938 . 
But https://github.com/Azure/azure-powershell/pull/3938 only cover the RMprofile!=null and SMProfile=null scenario, So this PR add all combination to avoid regression .
---

This checklist is used to make sure that common guidelines for a pull request are followed. You can find a more complete discussion of PowerShell cmdlet best practices [here](https://msdn.microsoft.com/en-us/library/dd878270(v=vs.85).aspx).

- [ Y] **I have read the [contribution guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md).**
- [Y ] **If changes were made to any cmdlet, the XML help was regenerated using the [platyPSHelp module](https://github.com/Azure/azure-powershell/blob/dev/documentation/platyPSHelp-documentation.md).**
- [ Y] **If any large changes are made to a service, they are reflected in the respective [change log](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#updating-the-change-log).**

### [General Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#general-guidelines)
- [ Y] Title of the pull request is clear and informative.
- [ Y] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [ ] The pull request does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/dev/documentation/breaking-changes.md) (unless a major version change occurs in the assembly and module).

### [Testing Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#testing-guidelines)
- [ Y] Pull request includes test coverage for the included changes.
- [ Y] PowerShell scripts used in tests should do any necessary setup as part of the test or suite setup, and should not use hard-coded values for locations or existing resources.

### [Cmdlet Signature Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#cmdlet-signature-guidelines)
- [ Y] New cmdlets that make changes or have side effects should implement `ShouldProcess` and have `SupportShouldProcess=true` specified in the cmdlet attribute. You can find more information on `ShouldProcess` [here](https://gist.github.com/markcowl/338e16fe5c8bbf195aff9f8af0db585d#what-is-the-change).
- [ Y] Cmdlet specifies `OutputType` attribute if any output is produced - if the cmdlet produces no output, it should implement a `PassThru` parameter.

### [Cmdlet Parameter Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#cmdlet-parameter-guidelines)
- [ Y] Parameter types should not expose types from the management library - complex parameter types should be defined in the module.
- [ Y] Complex parameter types are discouraged - a parameter type should be simple types as often as possible. If complex types are used, they should be shallow and easily creatable from a constructor or another cmdlet.
- [ Y] Cmdlet parameter sets should be mutually exclusive - each parameter set must have at least one mandatory parameter not in other parameter sets.

### CODE REVIEW

**Comment by blueww** [comment]
On Demand test: http://azuresdkci.cloudapp.net/view/1-AzurePowerShell/job/powershell-demand/1588/

File: src/Storage/Commands.Storage.Test/Common/Cmdlet/NewAzureStorageContextTest.cs [hunk 1]
```diff
@@ -20,9 +20,74 @@
 using Microsoft.WindowsAzure.Commands.Storage.Common.Cmdlet;
 using Microsoft.WindowsAzure.Storage;
 using Microsoft.WindowsAzure.Storage.Auth;
+using Microsoft.Azure.Commands.Common.Authentication;
+using Microsoft.Azure.Commands.Common.Authentication.Abstractions;
+using System.Collections.Generic;
 
 namespace Microsoft.WindowsAzure.Commands.Storage.Test.Common.Cmdlet
 {
+    class TestProfileProvider : AzureRmProfileProvider
```
Comment (markcowl) [inline src/Storage/Commands.Storage.Test/Common/Cmdlet/NewAzureStorageContextTest.cs]: @blueww Please add any additional scenarios to the tests I added, as xunit tests.  We are not accepting any new MSTest tests in the repo, and these need to be check-in tests

Comment (blueww) [inline src/Storage/Commands.Storage.Test/Common/Cmdlet/NewAzureStorageContextTest.cs]: Thanks for the comments!
Will try to find some time to fix it.

Comment (blueww) [inline src/Storage/Commands.Storage.Test/Common/Cmdlet/NewAzureStorageContextTest.cs]: The cases are added to the new test you added.
The PR is updated, please check.

**Comment by blueww** [comment]
New On Demand test: http://azuresdkci.cloudapp.net/view/1-AzurePowerShell/job/powershell-demand/1609/

File: src/Storage/Commands.Storage.Test/Common/Cmdlet/StorageContextDisconnectedTests.cs [hunk 2]
```diff
@@ -118,30 +152,148 @@ public void CanCreateStorageContextNameAndKey()
...
+                Assert.NotNull(storageContext);
+                Assert.Equal(cmdlet.StorageAccountName, storageContext.StorageAccountName);
+            }
+            finally
+            {
+                CleanupSession();
+            }
+        }
+
+        [Fact(Skip = "The case will fail since Environment is ignored when both SMProfile and RMProfile are null. Need confirm if this is a valid scenario.")]
```
Comment (markcowl) [inline src/Storage/Commands.Storage.Test/Common/Cmdlet/StorageContextDisconnectedTests.cs]: Please either remove this test, or file an issue for fixing the underlying issue and reference the issue id here.  The issue must be resolved by the end of the next sprint.

**Comment by markcowl** [comment]
@blueww one correction, otherwise LGTM

File: src/Storage/Commands.Storage.Test/Common/Cmdlet/StorageContextDisconnectedTests.cs (hunk 2 above)
Comment (blueww) [inline src/Storage/Commands.Storage.Test/Common/Cmdlet/StorageContextDisconnectedTests.cs]: The case is removed, and new on demand test: http://azuresdkci.cloudapp.net/view/1-AzurePowerShell/job/powershell-demand/1617/

We will continue to following this in mail to see if this is a valid scenario. If so, we need fix it.

**Comment by markcowl** [comment]
on demand run here: http://azuresdkci.cloudapp.net/view/1-AzurePowerShell/job/powershell-demand/1619/

**Pull request merged**